packages/runtime/src/interpreter/opcode-list.ts

```diff
@@ -2671,13 +2670,255 @@ export class ByteAdd extends Op {
     assertLen(args.length, 0, line);
   };
 
-  execute (stack: TEALStack): void {
+  execute (stack: TEALStack, op: MathOp): void {
     this.assertMinStackLen(stack, 2, this.line);
     const byteB = this.assertBytes(stack.pop(), this.line, MAX_INPUT_BYTE_LEN);
     const byteA = this.assertBytes(stack.pop(), this.line, MAX_INPUT_BYTE_LEN);
+    const bigIntB = bigEndianBytesToBigInt(byteB);
+    const bigIntA = bigEndianBytesToBigInt(byteA);
 
-    const result = bigEndianBytesToBigInt(byteA) + bigEndianBytesToBigInt(byteB);
-    const resultAsBytes = bigintToBigEndianBytes(result);
-    stack.push(this.assertBytes(resultAsBytes, this.line, MAX_OUTPUT_BYTE_LEN));
+    let r: bigint | boolean;
+    switch (op) {
+      case MathOp.Add: {
+        r = bigIntA + bigIntB;
+        break;
+      }
+      case MathOp.Sub: {
+        r = bigIntA - bigIntB;
+        this.checkUnderflow(r, this.line);
+        break;
+      }
+      case MathOp.Mul: {
+        r = bigIntA * bigIntB;
```

Comment:
this was discussed on call, we don't explicitly need a check here as max output length is 128 bytes (a * b). But since we already have it, let's keep it.